src/closure-types.ts

```diff
@@ -181,15 +184,61 @@ function convertUnion(
 
 function convertFunction(
     node: doctrine.type.FunctionType,
-    templateTypes: string[]): ts.FunctionType {
-  return new ts.FunctionType(
-      node.params.map(
-          (param, idx) => new ts.ParamType(
-              // TypeScript wants named parameters, but we don't have names.
-              'p' + idx,
-              convert(param, templateTypes))),
-      // Cast because typings are wrong: `FunctionType.result` is not an array.
-      node.result ? convert(node.result as any, templateTypes) : ts.anyType);
+    templateTypes: string[]): ts.FunctionType|ts.ConstructorType {
+  const params = node.params.map(
+      (param, idx) => new ts.ParamType(
+          // TypeScript wants named parameters, but we don't have names.
+          'p' + idx,
+          convert(param, templateTypes)));
+  if (node.new) {
+    return new ts.ConstructorType(
+        params,
+        // It doesn't make sense for a constructor to return something other
+        // than a named type. Also, in this context the name type is not
+        // nullable by default, so it's simpler to just directly convert here.
+        isName(node.this) ? new ts.NameType(node.this.name) : ts.anyType);
+  } else {
+    return new ts.FunctionType(
+        params,
+        // Cast because type is wrong: `FunctionType.result` is not an array.
+        node.result ? convert(node.result as any, templateTypes) : ts.anyType);
+  }
+}
+
+function convertRecord(node: doctrine.type.RecordType, templateTypes: string[]):
+    ts.RecordType|ts.NameType {
+  const fields = [];
+  for (const field of node.fields) {
+    if (field.type !== 'FieldType') {
+      return ts.anyType;
+    }
+    let fieldType =
+        field.value ? convert(field.value, templateTypes) : ts.anyType;
+
+    // In Closure you can't declare a record field optional, instead you
+    // declare `foo: bar|undefined`. In TypeScript we can represent this as
+    // `foo?: bar`. This also matches the semantics better, since Closure would
+    // allow the field to be omitted when it is `|undefined`, but TypeScript
+    // would require it to be explicitly set to `undefined`.
+    let optional = false;
+    if (fieldType.kind === 'union') {
+      fieldType.members = fieldType.members.filter((member) => {
+        if (member.kind === 'name' && member.name === 'undefined') {
+          optional = true;
+          return false;
+        }
+        return true;
+      });
+
+      // No need for a union if we collapsed it to one member.
+      if (fieldType.members.length === 1) {
```

Comment:
Isn't there code elsewhere that does this? Can it be unified with this code?

Reply:
Done, just calling simplify() now.